tsl/src/remote/dist_txn.c

```diff
@@ -177,36 +221,56 @@ dist_txn_xact_callback_1pc_pre_commit()
 		async_request_set_add(ars, remote_txn_async_send_commit(remote_txn));
 	}
 
-	testing_callback_call("waiting-commit");
+	eventcallback(DTXN_EVENT_WAIT_COMMIT);
 
 	/* async collect all the replies */
 	async_request_set_wait_all_ok_commands(ars);
 
 	dist_txn_deallocate_prepared_stmts_if_needed();
 }
 
-/* On abort on the frontend send aborts to all of the remote endpoints */
+/*
+ * Abort on the access node.
+ *
+ * The access node needs to send aborts to all of the remote endpoints. This
+ * code should not throw errors itself, since we are already in abort due to a
+ * previous error. Instead, we try to emit errors as warnings. For safety, we
+ * should probaby try-catch and swallow any potential lower-layer errors given
+ * that we're doing remote calls over the network. But the semantics for
+ * capturing and proceeding after such recursive errors are unclear.
+ */
 static void
 dist_txn_xact_callback_abort()
```

Comment:
An explicit ROLLBACK by the user, for instance, is not an error. It is a valid thing to do that doesn't necessarily disqualify the connection. For instance,

```
BEGIN;
INSERT INTO foo VALUES (....);
ROLLBACK;
```

This is not an error condition but still a rollback/abort. I think the WARNING is just to make clear that something is not right with the data node.

Reply:
No, they are not errors, but they cannot fail either, which is my concern, not the explicit rollback (or any rollback at all).

So, if the remote instance is just slow for some reason and we get a timeout, or the connection is broken, we will get spurious warnings in the log, which will just result in a flaky test (this can be handled, more discussions below).

At a more general level on aborts in distributed systems, if the user does an explicit rollback and, for some reason, the remote node does not respond, we have to guarantee an abort anyway (since aborts cannot fail). Closing the connection is a way to make sure that no spurious responses arrive late and also to quickly continue, but in this case, the transaction has still been rolled back implicitly on the remote and I do not think this warrant a warning.

Warnings are used when the user need to take special action to verify something, but in this case the transaction is still properly rolled back so I do not think this warrant a warning even though the rollback was done in a slightly strange way. It can be prudent to print a notice or info (which we then can filter out in the test above and prevent a flake).

The only case when I think that a warning or error is warranted is if something happens *before* the actual abort starts, e.g., a bad transaction id locally or some corrupt data structure.

Reply:
The connection is closed if abort fails. It just doesn't happen in this function. 

FWIW, this PR just refactors tests, it it is not intended to change the behavior of how we do aborts. And, I don't buy the argument that we should remove WARNINGs or filter output to avoid flaky tests; if we have that problem, then the behavior of the test is undefined (i.e., sometimes the rollback succeeds, sometimes it doesn't) and then the test itself is broken. Removing the WARNING will only hide the issue and is not a proper solution.

I guess we can discuss the log level, although I think WARNING seems appropriate for an something error-related (given that we cannot use ERROR).

Reply:
My main concern is that generated warnings are not very useful, but we can take that discussion separately instead of blocking this PR. I still think that we should not generate warnings on an abort expect if it is something local.